python/cudf/cudf/tests/dataset_generator.py

```diff
@@ -436,4 +489,86 @@ def timedelta_generator(dtype, size):
 
 
 def boolean_generator(size):
+    """
+    Generator for bool data
+    """
     return lambda: np.random.choice(a=[False, True], size=size)
+
+
+def get_nested_lists(dtype, nesting_depth, list_max_length):
```

Comment:
Man, it's hard to be sure if this is correct.
Here's what I did in C++, might be simpler to do something similar here:
Generate `num_rows` * `list_len`^`depth` elements
`depth` times:
  Split the list into a list of lists, where the "cut" offsets are generated incrementally using a distribution around the list_len parameter. You need `num_rows` * `list_len`^(`current_depth` - 1) offsets in each level (or until offset is > current length).
If this sounds like it could be easy to implement in Python, I would suggest it. AFAICT, the length distribution would be better. Otherwise we can try this one and see if it finds us some bugs.

Reply:
I spent a couple hours looking at this and couldn't come up with anything that was obviously better. FWIW I did come up with a recursive approach that I think gets the right answer:

```
top_level_list = []
def make_lists(list_max_len, nesting_depth, top_level_list):
    nesting_depth -= 1
    if nesting_depth > 0:
        L = np.random.randint(1, list_max_len)
        for i in range(L):
            top_level_list.append(make_lists(list_max_len, nesting_depth, []))
    else:
        top_level_list = np.random.randint(0, list_max_len, np.random.randint(1, list_max_len))
    return top_level_list
```

Reply:
This looks like a correct, straight-forward implementation. Thanks @brandon-b-miller !

Reply:
> This looks like a correct, straight-forward implementation. Thanks @brandon-b-miller !

+1

The recursive approach looks easier to maintain too, I plugged it into the existing helper functions and tested this locally, and verified for all different value types and works perfectly 👍

Reply:
@vuule @brandon-b-miller This is ready for another round of review.